File: frontend/services/SerialCommunication.ts
/**
 * USB Serial Communication Service
 * Direct communication with Jetson via telemetry radio
 * 
 * Replaces HTTP/WebSocket with efficient serial protocol
 */

/**
 * USB Serial Communication Service
 * Direct communication with Jetson via telemetry radio
 * 
 * Architecture:
 * Mobile (USB OTG) → Ground Radio (915MHz) → Air Radio → Jetson (/dev/ttyUSB0) → Backend
 */

import uuid from 'react-native-uuid';
import UsbSerialModule, { addDataListener } from './UsbSerialModule';

// ========== Message Types ==========

export interface SerialMessage {
  type: string;
  id?: string;
  timestamp: string;
  data?: any;
}

// Outgoing message types
export interface AgentCommandMessage extends SerialMessage {
  type: 'agent_command';
  data: {
    content: string;
  };
}

export interface ControlCommandMessage extends SerialMessage {
  type: 'control_command';
  data: {
    action: 'takeoff' | 'land' | 'emergency_stop' | 'velocity_body' | 'move_ned';
    altitude?: number;
    vx?: number;
    vy?: number;
    vz?: number;
    yaw_rate?: number;
    north?: number;
    east?: number;
    down?: number;
  };
}

export interface PlanConfirmMessage extends SerialMessage {
  type: 'plan_confirm';
  data: {
    plan_id: string;
  };
}

export interface PlanAbortMessage extends SerialMessage {
  type: 'plan_abort';
  data: {
    plan_id: string;
    feedback?: string;
  };
}

export interface StatusRequestMessage extends SerialMessage {
  type: 'status_request';
}

export interface ClearHistoryMessage extends SerialMessage {
  type: 'clear_history';
}

// Incoming message types
export interface AIMessageData {
  role: 'user' | 'assistant' | 'system';
  content: string;
  message_type?: 'info' | 'success' | 'warning' | 'error' | 'planning' | 'execution' | 'system';
  session_id?: string;
}

export interface PlanData {
  plan_id: string;
  summary: string;
  waypoints: Array<{
    id: string;
    description: string;
    action: string;
    position?: {
      lat: number;
      lon: number;
      alt: number;
    };
  }>;
  waypoint_count: number;
  estimated_duration?: number;
}

export interface TelemetryData {
  position?: {
    latitude: number;
    longitude: number;
    altitude_relative: number;
    altitude_absolute: number;
  };
  velocity?: {
    resultant: number;
  };
  heading?: number;
  battery?: {
    percentage: number;
    voltage: number;
    current?: number;
  };
  flight_mode?: string;
  armed?: boolean;
}

export interface BatteryData {
  percentage: number;
  voltage: number;
  current: number;
}

// Callbacks
type MessageCallback = (message: SerialMessage) => void;
type ConnectionCallback = (connected: boolean) => void;
type TelemetryCallback = (data: TelemetryData) => void;
type BatteryCallback = (data: BatteryData) => void;
type FlightModeCallback = (mode: string) => void;
type ArmedCallback = (armed: boolean) => void;
type AIMessageCallback = (message: any) => void;
type PlanConfirmationCallback = (plan: PlanData) => void;

// ========== Serial Communication Manager ==========

class SerialCommunicationManager {
  private deviceName: string | null = null;
  private connected = false;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private reconnectDelay = 2000;
  private lineBuffer = '';
  private dataListenerCleanup: (() => void) | null = null;

  // Callbacks
  private messageCallbacks: Set<MessageCallback> = new Set();
  private connectionCallbacks: Set<ConnectionCallback> = new Set();
  private telemetryCallbacks: Set<TelemetryCallback> = new Set();
  private batteryCallbacks: Set<BatteryCallback> = new Set();
  private flightModeCallbacks: Set<FlightModeCallback> = new Set();
  private armedCallbacks: Set<ArmedCallback> = new Set();
  private aiMessageCallbacks: Set<AIMessageCallback> = new Set();
  private planConfirmationCallbacks: Set<PlanConfirmationCallback> = new Set();

  // Pending responses (for request/response matching)
  private pendingResponses: Map<string, (response: SerialMessage) => void> = new Map();
  private responseTimeout = 10000; // 10 seconds

  /**
   * List available USB devices
   */
  async listDevices(): Promise<any[]> {
    try {
      return await UsbSerialModule.listDevices();
    } catch (error) {
      console.error('Failed to list USB devices:', error);
      return [];
    }
  }

  /**
   * Connect to USB serial device
   */
  async connect(deviceName?: string): Promise<boolean> {
    try {
      // If no device specified, use first available
      if (!deviceName) {
        const devices = await this.listDevices();
        console.log(`[CONNECT] Found ${devices.length} USB devices`);
        if (devices.length === 0) {
          console.warn('[CONNECT] No USB devices found');
          return false;
        }
        deviceName = devices[0].deviceName;
        console.log(`[CONNECT] Using device: ${deviceName}`);
      }

      console.log(`[CONNECT] Connecting to USB device: ${deviceName} @ 57600 baud`);
      
      const success = await UsbSerialModule.connect(deviceName, 57600);
      
      if (!success) {
        throw new Error('Failed to connect to device');
      }
      
      this.deviceName = deviceName;
      this.connected = true;
      this.lineBuffer = '';
      
      // Setup data listener
      this.dataListenerCleanup = addDataListener(this.handleSerialData);
      
      // Notify connection
      this.notifyConnectionStatus();
      
      console.log('[CONNECT] ✅ Serial connection established');
      return true;
    } catch (error) {
      console.error('Failed to connect to USB device:', error);
      this.connected = false;
      this.notifyConnectionStatus();
      this.scheduleReconnect();
      return false;
    }
  }

  /**
   * Handle incoming serial data
   */
  private handleSerialData = (data: string): void => {
    // Append to line buffer
    this.lineBuffer += data;
    console.log(`[RX] Received ${data.length} chars, buffer: ${this.lineBuffer.length}`);

    // Process complete lines (newline-delimited)
    const lines = this.lineBuffer.split('\n');
    
    // Keep incomplete line in buffer
    this.lineBuffer = lines.pop() || '';

    // Process complete lines
    lines.forEach(line => {
      const trimmed = line.trim();
      if (trimmed) {
        try {
          const message: SerialMessage = JSON.parse(trimmed);
          this.handleMessage(message);
        } catch (error) {
          console.error('Failed to parse serial message:', trimmed, error);
        }
      }
    });
  };

  /**
   * Handle parsed message
   */
  private handleMessage(message: SerialMessage): void {
    // Check for pending response
    if (message.id) {
      const callback = this.pendingResponses.get(message.id);
      if (callback) {
        callback(message);
        this.pendingResponses.delete(message.id);
        return; // Don't process further
      }
    }

    // Route by message type
    switch (message.type) {
      case 'telemetry':
        if (message.data) {
          this.notifyTelemetryCallbacks(message.data);
        }
        break;

      case 'battery':
        if (message.data) {
          this.notifyBatteryCallbacks(message.data);
        }
        break;

      case 'flight_mode':
        if (message.data?.mode) {
          this.notifyFlightModeCallbacks(message.data.mode);
        }
        break;

      case 'armed':
        if (message.data?.armed !== undefined) {
          this.notifyArmedCallbacks(message.data.armed);
        }
        break;

      case 'ai_message':
        // AI messages from agent
        this.notifyAIMessageCallbacks(message);
        break;

      case 'plan_confirmation':
        // Plan confirmation requests
        if (message.data) {
          this.notifyPlanConfirmationCallbacks(message.data);
        }
        break;

      default:
        // Generic message callback (for AI messages, errors, etc.)
        this.notifyMessageCallbacks(message);
        break;
    }
  }

  /**
   * Send message to backend via serial (with chunking for radio stability)
   */
  async sendMessage(message: Omit<SerialMessage, 'timestamp'>): Promise<boolean> {
    console.log(`[SEND_MESSAGE] Step 1: Called with type=${message.type}, connected=${this.connected}`);
    
    if (!this.connected) {
      console.warn('[SEND_MESSAGE] Blocked: Serial not connected');
      return false;
    }

    try {
      console.log(`[SEND_MESSAGE] Step 2: Connection check passed`);
      
      // Add timestamp
      const fullMessage: SerialMessage = {
        ...message,
        timestamp: new Date().toISOString(),
      };
      console.log(`[SEND_MESSAGE] Step 3: Timestamp added`);

      // Serialize to JSON and add newline
      const data = JSON.stringify(fullMessage) + '\n';
      console.log(`[TX] Step 4: Serialized ${data.length} bytes: ${data.substring(0, 80)}...`);
      
      // CRITICAL: Send in small chunks with delays for radio stability
      // Radio buffer is small (~64-128 bytes), sending all at once causes corruption
      const CHUNK_SIZE = 16; // Send 16 bytes at a time
      const CHUNK_DELAY_MS = 30; // 30ms delay between chunks
      
      console.log(`[TX] Step 5: Sending in ${CHUNK_SIZE}-byte chunks with ${CHUNK_DELAY_MS}ms delays...`);
      
      const totalChunks = Math.ceil(data.length / CHUNK_SIZE);
      for (let i = 0; i < data.length; i += CHUNK_SIZE) {
        const chunk = data.substring(i, i + CHUNK_SIZE);
        const chunkNum = Math.floor(i / CHUNK_SIZE) + 1;
        
        console.log(`[TX] Chunk ${chunkNum}/${totalChunks}: ${chunk.length} bytes`);
        
        const success = await UsbSerialModule.write(chunk);
        if (!success) {
          console.error(`[TX] ❌ Chunk ${chunkNum} failed`);
          return false;
        }
        
        // Wait between chunks (let radio transmit)
        if (i + CHUNK_SIZE < data.length) {
          await new Promise(resolve => setTimeout(resolve, CHUNK_DELAY_MS));
        }
      }
      
      console.log('[TX] ✅ All chunks sent successfully');
      return true;
      
    } catch (error) {
      console.error('[SEND_MESSAGE] Exception in sendMessage:');
      console.error('[SEND_MESSAGE] Error type:', typeof error);
      console.error('[SEND_MESSAGE] Error:', error);
      if (error && typeof error === 'object') {
        console.error('[SEND_MESSAGE] Error keys:', Object.keys(error));
        console.error('[SEND_MESSAGE] Error.message:', (error as any).message);
        console.error('[SEND_MESSAGE] Error.code:', (error as any).code);
      }
      this.disconnect();
      return false;
    }
  }

  /**
   * Send message and wait for response
   */
  async sendWithResponse(
    message: Omit<SerialMessage, 'id' | 'timestamp'>,
    timeout: number = this.responseTimeout
  ): Promise<SerialMessage> {
    const id = uuid.v4() as string;
    const fullMessage = { ...message, id };

    return new Promise(async (resolve, reject) => {
      // Setup timeout
      const timer = setTimeout(() => {
        this.pendingResponses.delete(id);
        reject(new Error('Response timeout'));
      }, timeout);

      // Setup response callback
      this.pendingResponses.set(id, (response) => {
        clearTimeout(timer);
        resolve(response);
      });

      // Send message
      const success = await this.sendMessage(fullMessage);
      if (!success) {
        clearTimeout(timer);
        this.pendingResponses.delete(id);
        reject(new Error('Failed to send message'));
      }
    });
  }

  /**
   * Send agent command (AI chat)
   */
  async sendAgentCommand(content: string): Promise<boolean> {
    return this.sendMessage({
      type: 'agent_command',
      id: uuid.v4() as string,
      data: { content },
    });
  }

  /**
   * Send control command
   */
  async sendControlCommand(
    action: 'takeoff' | 'land' | 'emergency_stop' | 'velocity_body' | 'move_ned',
    params?: any
  ): Promise<boolean> {
    try {
      console.log(`[SEND_CONTROL] Step 1: Action=${action}, Connected=${this.connected}`);
      
      let msgId: string;
      try {
        msgId = uuid.v4() as string;
        console.log(`[SEND_CONTROL] Step 2: UUID generated = ${msgId}`);
      } catch (uuidError) {
        console.error(`[SEND_CONTROL] UUID generation failed:`, uuidError);
        throw uuidError;
      }
      
      const message = {
        type: 'control_command',
        id: msgId,
        data: {
          action,
          ...params,
        },
      };
      console.log(`[SEND_CONTROL] Step 3: Message created`);
      
      console.log(`[SEND_CONTROL] Step 4: Calling sendMessage...`);
      const result = await this.sendMessage(message);
      console.log(`[SEND_CONTROL] Step 5: sendMessage returned ${result}`);
      return result;
    } catch (error) {
      console.error(`[SEND_CONTROL] Exception caught:`, error);
      console.error(`[SEND_CONTROL] Error type: ${typeof error}`);
      console.error(`[SEND_CONTROL] Error keys: ${Object.keys(error || {})}`);
      if (error && typeof error === 'object') {
        console.error(`[SEND_CONTROL] Error.message: ${(error as any).message}`);
        console.error(`[SEND_CONTROL] Error.toString: ${error.toString()}`);
      }
      return false;
    }
  }

  /**
   * Confirm plan (same as confirmPlan but for consistency)
   */
  async sendPlanConfirm(planId: string): Promise<boolean> {
    return this.confirmPlan(planId);
  }

  /**
   * Abort plan (same as abortPlan but for consistency)
   */
  async sendPlanAbort(planId: string, feedback?: string): Promise<boolean> {
    return this.abortPlan(planId, feedback);
  }

  /**
   * Clear history (same as clearHistory but for consistency)
   */
  async sendClearHistory(): Promise<boolean> {
    return this.clearHistory();
  }

  /**
   * Confirm plan
   */
  async confirmPlan(planId: string): Promise<boolean> {
    return this.sendMessage({
      type: 'plan_confirm',
      id: uuid.v4() as string,
      data: { plan_id: planId },
    });
  }

  /**
   * Abort plan
   */
  async abortPlan(planId: string, feedback?: string): Promise<boolean> {
    return this.sendMessage({
      type: 'plan_abort',
      id: uuid.v4() as string,
      data: { plan_id: planId, feedback },
    });
  }

  /**
   * Request status
   */
  async requestStatus(): Promise<SerialMessage> {
    return this.sendWithResponse({
      type: 'status_request',
    });
  }

  /**
   * Clear history
   */
  async clearHistory(): Promise<boolean> {
    return this.sendMessage({
      type: 'clear_history',
      id: uuid.v4() as string,
    });
  }

  /**
   * Emergency stop
   */
  async emergencyStop(): Promise<SerialMessage> {
    return this.sendWithResponse({
      type: 'control_command',
      data: { action: 'emergency_stop' },
    }, 5000); // Shorter timeout for emergency
  }

  // ========== Subscription Methods ==========

  onMessage(callback: MessageCallback): () => void {
    this.messageCallbacks.add(callback);
    return () => this.messageCallbacks.delete(callback);
  }

  onConnectionChange(callback: ConnectionCallback): () => void {
    callback(this.connected); // Immediate notification
    this.connectionCallbacks.add(callback);
    return () => this.connectionCallbacks.delete(callback);
  }

  onTelemetry(callback: TelemetryCallback): () => void {
    this.telemetryCallbacks.add(callback);
    return () => this.telemetryCallbacks.delete(callback);
  }

  onBattery(callback: BatteryCallback): () => void {
    this.batteryCallbacks.add(callback);
    return () => this.batteryCallbacks.delete(callback);
  }

  onFlightMode(callback: FlightModeCallback): () => void {
    this.flightModeCallbacks.add(callback);
    return () => this.flightModeCallbacks.delete(callback);
  }

  onArmed(callback: ArmedCallback): () => void {
    this.armedCallbacks.add(callback);
    return () => this.armedCallbacks.delete(callback);
  }

  onAIMessage(callback: AIMessageCallback): () => void {
    this.aiMessageCallbacks.add(callback);
    return () => this.aiMessageCallbacks.delete(callback);
  }

  onPlanConfirmation(callback: PlanConfirmationCallback): () => void {
    this.planConfirmationCallbacks.add(callback);
    return () => this.planConfirmationCallbacks.delete(callback);
  }

  // ========== Notification Methods ==========

  private notifyMessageCallbacks(message: SerialMessage): void {
    this.messageCallbacks.forEach(callback => {
      try {
        callback(message);
      } catch (error) {
        console.error('Message callback error:', error);
      }
    });
  }

  private notifyConnectionStatus(): void {
    this.connectionCallbacks.forEach(callback => {
      try {
        callback(this.connected);
      } catch (error) {
        console.error('Connection callback error:', error);
      }
    });
  }

  private notifyTelemetryCallbacks(data: TelemetryData): void {
    this.telemetryCallbacks.forEach(callback => {
      try {
        callback(data);
      } catch (error) {
        console.error('Telemetry callback error:', error);
      }
    });
  }

  private notifyBatteryCallbacks(data: BatteryData): void {
    this.batteryCallbacks.forEach(callback => {
      try {
        callback(data);
      } catch (error) {
        console.error('Battery callback error:', error);
      }
    });
  }

  private notifyFlightModeCallbacks(mode: string): void {
    this.flightModeCallbacks.forEach(callback => {
      try {
        callback(mode);
      } catch (error) {
        console.error('Flight mode callback error:', error);
      }
    });
  }

  private notifyArmedCallbacks(armed: boolean): void {
    this.armedCallbacks.forEach(callback => {
      try {
        callback(armed);
      } catch (error) {
        console.error('Armed callback error:', error);
      }
    });
  }

  private notifyAIMessageCallbacks(message: SerialMessage): void {
    this.aiMessageCallbacks.forEach(callback => {
      try {
        callback(message);
      } catch (error) {
        console.error('AI message callback error:', error);
      }
    });
  }

  private notifyPlanConfirmationCallbacks(plan: PlanData): void {
    this.planConfirmationCallbacks.forEach(callback => {
      try {
        callback(plan);
      } catch (error) {
        console.error('Plan confirmation callback error:', error);
      }
    });
  }

  // ========== Connection Management ==========

  /**
   * Schedule reconnection
   */
  private scheduleReconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
    }

    this.reconnectTimer = setTimeout(() => {
      console.log('Attempting to reconnect serial...');
      this.connect(this.deviceName || undefined);
    }, this.reconnectDelay) as unknown as NodeJS.Timeout;
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.connected;
  }

  /**
   * Disconnect
   */
  async disconnect(): Promise<void> {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.dataListenerCleanup) {
      this.dataListenerCleanup();
      this.dataListenerCleanup = null;
    }

    try {
      await UsbSerialModule.disconnect();
    } catch (error) {
      console.error('Error during disconnect:', error);
    }

    this.deviceName = null;
    this.connected = false;
    this.lineBuffer = '';
    this.pendingResponses.clear();
    this.notifyConnectionStatus();
  }
}

// Singleton instance
export const serialComm = new SerialCommunicationManager();

File: backend/services/serial_communication.py
"""
Serial Communication Service for Backend
Handles USB serial communication with mobile app via telemetry radio

Replaces HTTP/WebSocket endpoints with direct serial protocol
"""

import asyncio
import json
import logging
import serial
import uuid
from datetime import datetime
from typing import Optional, Dict, Any, Callable
from collections import deque

logger = logging.getLogger("serial_comm")

# Message queue for outgoing messages
OutgoingQueue = deque[Dict[str, Any]]


class SerialCommunicationManager:
    """
    Manages serial communication with mobile app
    """

    def __init__(
        self,
        port: str = "/dev/ttyUSB0",
        baudrate: int = 115200,
        timeout: float = 0.1,
    ):
        self.port = port
        self.baudrate = baudrate
        self.timeout = timeout
        self.serial: Optional[serial.Serial] = None
        self.connected = False
        self.running = False

        # Message handlers
        self.message_handlers: Dict[str, Callable] = {}

        # Outgoing message queue
        self.outgoing_queue: OutgoingQueue = deque(maxlen=100)

        # Line buffer for incoming data
        self.line_buffer = ""

    def register_handler(self, message_type: str, handler: Callable) -> None:
        """
        Register a handler for a specific message type

        Args:
            message_type: Type of message to handle
            handler: Async function to handle the message
        """
        self.message_handlers[message_type] = handler
        logger.info(f"Registered handler for message type: {message_type}")

    async def connect(self) -> bool:
        """
        Connect to serial port

        Returns:
            True if connected successfully
        """
        try:
            logger.info(f"Connecting to serial port {self.port} @ {self.baudrate}")
            
            self.serial = serial.Serial(
                port=self.port,
                baudrate=self.baudrate,
                timeout=self.timeout,
                write_timeout=1.0,
            )

            self.connected = True
            logger.info("Serial connection established")
            return True

        except Exception as e:
            logger.error(f"Failed to connect to serial port: {e}")
            self.connected = False
            return False

    async def disconnect(self) -> None:
        """Close serial connection"""
        self.running = False

        if self.serial and self.serial.is_open:
            try:
                self.serial.close()
                logger.info("Serial connection closed")
            except Exception as e:
                logger.error(f"Error closing serial port: {e}")

        self.serial = None
        self.connected = False

    async def send_message(self, message: Dict[str, Any]) -> bool:
        """
        Send message to mobile app

        Args:
            message: Message dictionary

        Returns:
            True if sent successfully
        """
        if not self.connected or not self.serial:
            logger.warning("Serial not connected, cannot send message")
            return False

        try:
            # Add timestamp if not present
            if "timestamp" not in message:
                message["timestamp"] = datetime.utcnow().isoformat() + "Z"

            # Serialize to JSON and add newline
            data = json.dumps(message) + "\n"

            # Write to serial
            self.serial.write(data.encode("utf-8"))
            self.serial.flush()

            return True

        except Exception as e:
            logger.error(f"Failed to send serial message: {e}")
            return False

    def queue_message(self, message: Dict[str, Any]) -> None:
        """
        Add message to outgoing queue

        Args:
            message: Message to send
        """
        self.outgoing_queue.append(message)

    async def send_response(
        self,
        request_id: str,
        message_type: str,
        data: Optional[Dict[str, Any]] = None,
    ) -> bool:
        """
        Send response to a request

        Args:
            request_id: ID of the original request
            message_type: Type of response message
            data: Response data

        Returns:
            True if sent successfully
        """
        message = {
            "type": message_type,
            "id": request_id,
            "timestamp": datetime.utcnow().isoformat() + "Z",
        }

        if data:
            message["data"] = data

        return await self.send_message(message)

    async def send_telemetry(
        self,
        telemetry_type: str,
        data: Dict[str, Any],
    ) -> bool:
        """
        Send telemetry update (no request ID)

        Args:
            telemetry_type: Type of telemetry (telemetry, battery, flight_mode, armed)
            data: Telemetry data

        Returns:
            True if sent successfully
        """
        message = {
            "type": telemetry_type,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "data": data,
        }

        return await self.send_message(message)

    async def send_error(
        self,
        request_id: Optional[str],
        error: str,
        code: str = "ERROR",
    ) -> bool:
        """
        Send error message

        Args:
            request_id: ID of the request that caused error (if any)
            error: Error message
            code: Error code

        Returns:
            True if sent successfully
        """
        message = {
            "type": "error",
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "data": {
                "error": error,
                "code": code,
            },
        }

        if request_id:
            message["id"] = request_id

        return await self.send_message(message)

    async def read_message(self) -> Optional[Dict[str, Any]]:
        """
        Read and parse a message from serial with robust buffering
        Handles fragmented messages from 915MHz radio transmission

        Returns:
            Parsed message dict or None
        """
        if not self.connected or not self.serial:
            return None

        try:
            # Read available data
            if self.serial.in_waiting > 0:
                data = self.serial.read(self.serial.in_waiting).decode("utf-8", errors='ignore')
                self.line_buffer += data
                logger.debug(f"[SERIAL RX] Received {len(data)} bytes, buffer now {len(self.line_buffer)} bytes")
                
                # CRITICAL: Wait 150ms for radio to transmit all chunks
                # Radio sends messages in 2-5 packets with 20-50ms delays
                await asyncio.sleep(0.15)
                
                # Read additional packets that arrived during wait (up to 5 rounds)
                additional_rounds = 0
                while self.serial.in_waiting > 0 and additional_rounds < 5:
                    data = self.serial.read(self.serial.in_waiting).decode("utf-8", errors='ignore')
                    self.line_buffer += data
                    additional_rounds += 1
                    logger.debug(f"[SERIAL RX] Additional packet #{additional_rounds}: {len(data)} bytes, buffer now {len(self.line_buffer)} bytes")
                    
                    # Wait a bit more for stragglers
                    await asyncio.sleep(0.05)

            # Process complete lines (newline-delimited JSON)
            if "\n" in self.line_buffer:
                lines = self.line_buffer.split("\n")
                self.line_buffer = lines[-1]  # Keep incomplete line in buffer

                # Process all complete lines
                for line in lines[:-1]:
                    line = line.strip()
                    if not line:
                        continue
                    
                    logger.info(f"[SERIAL RX] Processing line ({len(line)} chars): {line[:100]}...")
                    
                    # Try to parse directly first
                    try:
                        message = json.loads(line)
                        logger.info(f"[SERIAL RX] ✅ Parsed message type: {message.get('type')}")
                        return message
                    except json.JSONDecodeError as e:
                        # If direct parse fails, try recovery
                        logger.warning(f"[SERIAL RX] Parse failed, attempting recovery...")
                        
                        # Try to find valid JSON object in the line
                        best_json = None
                        for start_idx in range(len(line)):
                            if line[start_idx] == "{":
                                # Count braces from this starting point
                                brace_count = 0
                                for end_idx in range(start_idx, len(line)):
                                    if line[end_idx] == "{":
                                        brace_count += 1
                                    elif line[end_idx] == "}":
                                        brace_count -= 1
                                        if brace_count == 0:
                                            # Found matching closing brace
                                            candidate = line[start_idx:end_idx+1]
                                            # Try to parse to validate
                                            try:
                                                test = json.loads(candidate)
                                                # Check if it has required fields
                                                if 'type' in test or 'data' in test or 'action' in test:
                                                    best_json = candidate
                                                    logger.info(f"[SERIAL RX] ✅ Recovered valid message")
                                                    break
                                            except:
                                                pass
                                if best_json:
                                    break
                        
                        if best_json:
                            try:
                                message = json.loads(best_json)
                                logger.info(f"[SERIAL RX] ✅ Parsed recovered message type: {message.get('type')}")
                                return message
                            except:
                                pass
                        
                        logger.error(f"[SERIAL RX] ❌ Cannot recover valid JSON from: {line[:200]}")

            return None

        except Exception as e:
            logger.error(f"Error reading from serial: {e}")
            return None

    async def handle_message(self, message: Dict[str, Any]) -> None:
        """
        Handle incoming message with fallback for fragmented messages

        Args:
            message: Parsed message dictionary
        """
        message_type = message.get("type")
        message_id = message.get("id")

        logger.info(f"[HANDLER] Processing message type: {message_type}, id: {message_id}")

        # FALLBACK: If no "type" but has "action", treat as control_command
        # This handles fragmented messages from radio where beginning is lost
        if not message_type and "action" in message:
            logger.info(f"[HANDLER] No type field, but found 'action' - treating as control_command")
            message_type = "control_command"
            # Wrap action in data field if not already there
            if "data" not in message:
                message["data"] = {
                    "action": message.get("action"),
                    "altitude": message.get("altitude", 2.5)
                }

        if not message_type:
            logger.warning("Received message without type and no action field")
            await self.send_error(message_id, "Missing message type", "INVALID_MESSAGE")
            return

        # Get handler for this message type
        handler = self.message_handlers.get(message_type)

        if not handler:
            logger.warning(f"No handler for message type: {message_type}")
            await self.send_error(
                message_id, f"Unknown message type: {message_type}", "UNKNOWN_TYPE"
            )
            return

        try:
            # Call handler
            logger.info(f"[HANDLER] Calling handler for {message_type}")
            await handler(message)
            logger.info(f"[HANDLER] Handler completed for {message_type}")

        except Exception as e:
            logger.error(f"Error handling message {message_type}: {e}", exc_info=True)
            await self.send_error(
                message_id, f"Internal error: {str(e)}", "INTERNAL_ERROR"
            )

    async def run(self) -> None:
        """
        Main loop for reading/writing serial messages
        """
        self.running = True

        logger.info("Starting serial communication loop")

        while self.running:
            try:
                # Read incoming message
                message = await self.read_message()
                if message:
                    await self.handle_message(message)

                # Send queued messages
                while self.outgoing_queue:
                    outgoing = self.outgoing_queue.popleft()
                    await self.send_message(outgoing)

                # Small delay to prevent CPU spinning
                await asyncio.sleep(0.01)

            except Exception as e:
                logger.error(f"Error in serial communication loop: {e}", exc_info=True)
                await asyncio.sleep(1.0)

        logger.info("Serial communication loop stopped")


# Global singleton instance
_serial_comm: Optional[SerialCommunicationManager] = None


def get_serial_comm() -> SerialCommunicationManager:
    """Get the global serial communication manager"""
    global _serial_comm
    if _serial_comm is None:
        # Get port from config
        from config import SERIAL_PORT, SERIAL_BAUDRATE

        _serial_comm = SerialCommunicationManager(
            port=SERIAL_PORT,
            baudrate=SERIAL_BAUDRATE,
        )
    return _serial_comm


async def init_serial_communication() -> SerialCommunicationManager:
    """
    Initialize serial communication

    Returns:
        SerialCommunicationManager instance
    """
    serial_comm = get_serial_comm()

    # Connect to serial port
    success = await serial_comm.connect()
    if not success:
        logger.error("Failed to initialize serial communication")
        raise RuntimeError("Failed to connect to serial port")

    logger.info("Serial communication initialized successfully")
    return serial_comm